src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -8370,30 +8370,50 @@ private StatementSyntax ParseLocalDeclarationStatement()
             }
         }
 
-        private VariableDesignationSyntax ParseDesignation()
+        private VariableDesignationSyntax ParseDesignation(bool forPattern)
         {
             // the two forms of designation are
             // (1) identifier
             // (2) ( designation ... )
+            // for pattern-matching, we permite the designation list to be empty
             VariableDesignationSyntax result;
             if (this.CurrentToken.Kind == SyntaxKind.OpenParenToken)
             {
                 var openParen = this.EatToken(SyntaxKind.OpenParenToken);
                 var listOfDesignations = _pool.AllocateSeparated<VariableDesignationSyntax>();
 
-                listOfDesignations.Add(ParseDesignation());
-                listOfDesignations.AddSeparator(this.EatToken(SyntaxKind.CommaToken));
-
-                while (true)
+                if (forPattern)
                 {
-                    listOfDesignations.Add(ParseDesignation());
-                    if (this.CurrentToken.Kind == SyntaxKind.CommaToken)
+                    // for pattern-matching, we permit a designation list that contains zero or one element
+                    while (this.CurrentToken.Kind != SyntaxKind.CloseParenToken)
                     {
-                        listOfDesignations.AddSeparator(this.EatToken(SyntaxKind.CommaToken));
+                        listOfDesignations.Add(ParseDesignation(forPattern));
+                        if (this.CurrentToken.Kind == SyntaxKind.CommaToken)
+                        {
+                            listOfDesignations.AddSeparator(this.EatToken(SyntaxKind.CommaToken));
+                        }
+                        else
+                        {
+                            break;
+                        }
                     }
-                    else
+                }
+                else
+                {
+                    listOfDesignations.Add(ParseDesignation(forPattern));
+                    listOfDesignations.AddSeparator(this.EatToken(SyntaxKind.CommaToken));
+
+                    while (true)
```

Comment:
>while (true) [](start = 20, length = 12)

Would it make sense to share the `while` loop between the two cases? Perhaps something like:
```c#
bool done = false;
if (forPattern)
{
    done = (this.CurrentToken.Kind == SyntaxKind.CloseParenToken);
}
else
{
    listOfDesignations.Add(ParseDesignation(forPattern));
    listOfDesignations.AddSeparator(EatToken(SyntaxKind.CommaToken));
}
if (!done)
{
    while (true)
    {
        // ...
    }
}
```